### PR TITLE
CT-125 Added casa 4.3.0 to centos6. Also added support for Admit, inc…

### PR DESCRIPTION
**Pull request by yeunga**
…luding Sphinx and dot(Graphviz). Added patch to address the leap second issue (issue 33).

### CODE REVIEW

**Comment by brianmajor** [comment]
The changes look good so far Alinga

File: software-containers/README.md [hunk 1]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
 ### xterm
 `xterm` must be installed on the container
+## Creating a CASA container
+This section describes the process of creating a CASA container. The currently supported CASA containers are in the arcade-casa folder. The process iterates over build, deploy and test.
+### Building A Container
+To build a container, go into the relevant folder with the build files or create a new folder if necessary. There are two main files to building a container: Makefile and Dockerfile.
+#### Makefile
+This is the main file that controls the build process. This file is updated to add/change/remove a CASA version or to change the build process. It invokes docker to build the container.
+#### Dockerfile
+This file controls the building of the CASA container. For example, for versions 4.3.x to 4.4.x, the ADMIT package is not added to the container. Changes to the container are made in this file.
```
Comment (brianmajor) [inline software-containers/README.md]: I think you should omit these three sections (building a container, makefile, dockerfile) and just keep brief instructions on building in your 'how to build' section.  there's too much repitition.

File: software-containers/README.md [hunk 2]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
+#### How To Build
+1. cd into the relevant folder with the build file or create a new folder if necessary
+2. update or create the relevant files, e.g. Makefile, Dockerfile, download.sh
+3. execute 'make'
+4. execute 'docker image ls'
```
Comment (brianmajor) [inline software-containers/README.md]: Step 4 isn't needed so please omit it and the rest of the text following in this section.

File: software-containers/README.md [hunk 3]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
+
+Step (3) generates a docker image in the local machine, which is listed in step(4). Here is an example list of generated docker images:
+| REPOSITORY | TAG | IMAGE ID | CREATED | SIZE |
+| --------------------- | ------- | ---------------- | --------------- | ------- |
+| bucket.canfar.net/arcade-casa | 4.5.3-el6 | 0b8c5890921b | 13 hours ago | 4.72GB |
+| bucket.canfar.net/arcade-casa | 4.5.2-el6 | 21719143194a | 13 hours ago | 4.72GB |
+
+For work that does not require a UI, for example adding a new package to CASA, we may be able to run the container and test our work at this point. Execute 'docker run -it <docker image ID> /bin/bash' and check if the work is correct.
+### Deploying A Container
+Since a CASA container runs in Arcade under Kubernetes, the container image generated from the build step needs to be pushed to the repository that Kubernetes uses.
```
Comment (brianmajor) [inline software-containers/README.md]: The steps in "Deploying a container" are for CADC only so should not be described here.  They're good for your own notes but not the right information for open-source collaborators.

File: software-containers/README.md [hunk 4]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
+CASA <3>: exit
+--------> exit()
+leaving casa...
+
+cwd: ~
+ > casaviewer
+
+cwd: ~
+```
+### Updating The Config Map
```
Comment (brianmajor) [inline software-containers/README.md]: Please omit this section--it is CADC specific.

File: software-containers/README.md [hunk 5]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
+```
+#!/bin/bash
+
+IP_ADDRESS=`hostname --ip-address`
+
+/opt/shibboleth/bin/curl -v -L -k -E /cavern/home/${arcade_username}/.ssl/cadcproxy.pem -d "target-ip=${IP_ADDRESS}" -d "software=canfar-registry.cloud.computecanada.ca/arcade-casa:4.7.0-el7" https://${arcade_hostname}/arcade/session/${VNC_PW}/app
+```
+##### Use The UI
+If you have previously created a desktop session, you can use the UI of that session to access your container. There is an 'Applications' tab on the top left hand corner. Select 'Applications->CANFAR->CASA' and then the version under development. An xterm for the selected CASA version will pop up after some time.
+#### Perform Tests
```
Comment (brianmajor) [inline software-containers/README.md]: This information in 'perform tests' is good so you can keep this.

File: software-containers/README.md [hunk 6]
```diff
@@ -16,3 +16,156 @@ group:      sss files
...
+```
+5. tag the image with the repository that Kubernetes uses
+```
+docker tag <docker image ID> canfar-registry.cloud.computecanada.ca/arcade-casa:<TAG>
+```
+6. push the image to the repository that Kubernetes uses
+```
+docker push canfar-registry.cloud.computecanada.ca/arcade-casa:<TAG>
+```
+### Testing A Container
```
Comment (brianmajor) [inline software-containers/README.md]: In this section please omit everything except for the info in "perform tests"

**Review by brianmajor — CHANGES_REQUESTED** [review]
Everything looks good except for the content in README.md.  The document is great for internal notes but are not applicable to an open-source audience.  Different groups who wish to use these containers will have a different way of deploying them, so all deployment details should be omitted.

File: software-containers/README.md [hunk 7]
```diff
@@ -18,65 +18,13 @@ group:      sss files
...
-Execute the following command and open a browser using the desktop session URL returned
-```
-curl -E ~/.ssl/cadcproxy.pem https://proto.canfar.net/arcade/session -d "name=mydesktop" -d "type=desktop"
-```
-#### Get Access The CASA Container
+Since a CASA container resides inside Arcade, we need to access an Arcade desktop session. 
+#### Getring Access To The CASA Container
 ##### Use A Custom Script
 We can create a custom script to access the container. We can save the script in our home directory, which persists over sessions. To access the CASA container, we just execute the script. The following is an example of such a script. Just replace the version number (4.7.0-el7) with the version under test. An xterm for the CASA version will pop up after some time.
 ```
```
Comment (brianmajor) [inline software-containers/README.md]: Please delete "use a custom script" section

File: software-containers/README.md [hunk 8]
```diff
@@ -88,7 +36,7 @@ IP_ADDRESS=`hostname --ip-address`
...
 ##### Use The UI
 If you have previously created a desktop session, you can use the UI of that session to access your container. There is an 'Applications' tab on the top left hand corner. Select 'Applications->CANFAR->CASA' and then the version under development. An xterm for the selected CASA version will pop up after some time.
```
Comment (brianmajor) [inline software-containers/README.md]: You can repurpose this section to indicate that "you will need to run the container in a graphical environment, such as arcade, to test the visual components of the CASA containers".  But no other details about Arcade please.  Those things will likely all change and we don't want to have to come back and update this.

File: software-containers/README.md [hunk 9]
```diff
@@ -18,65 +18,13 @@ group:      sss files
 `xterm` must be installed on the container
 ## Creating a CASA container
```
Comment (brianmajor) [inline software-containers/README.md]: You can change this header to be "Building a CASA container" and then only have the steps.

**Review by brianmajor — CHANGES_REQUESTED** [review]
This documentation should be in the arcade-casa directory, not in software-containers.  Please restore the README.md for software containers and move this into arcade-casa.

File: software-containers/README.md (hunk 6 above)
Comment (yeunga) [reply software-containers/README.md]: Done

File: software-containers/README.md (hunk 5 above)
Comment (yeunga) [reply software-containers/README.md]: Okay

File: software-containers/README.md (hunk 7 above)
Comment (yeunga) [reply software-containers/README.md]: Okay.

File: software-containers/README.md (hunk 8 above)
Comment (yeunga) [reply software-containers/README.md]: Okay

File: software-containers/README.md (hunk 3 above)
Comment (yeunga) [reply software-containers/README.md]: Done

File: software-containers/README.md (hunk 2 above)
Comment (yeunga) [reply software-containers/README.md]: Done

File: software-containers/README.md (hunk 1 above)
Comment (yeunga) [reply software-containers/README.md]: Done

File: software-containers/README.md (hunk 4 above)
Comment (yeunga) [reply software-containers/README.md]: Done

File: software-containers/README.md (hunk 9 above)
Comment (yeunga) [reply software-containers/README.md]: Done.

File: software-containers/arcade-casa/README.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
+# Arcade CASA Container
+## Building A CASA container
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]: A -> a

File: software-containers/arcade-casa/README.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+1. cd into the relevant folder with the build file or create a new folder if necessary
+2. update or create the relevant files, e.g. Makefile, Dockerfile, download.sh
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]: I'm not sure what point 2 means here.. are you talking about building what is in git or extending it for new versions?

File: software-containers/arcade-casa/README.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+3. execute 'make'
+### Testing A Container
+#### Getting Access To The CASA Container
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]: Remove this title

File: software-containers/arcade-casa/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+##### Using The UI
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]: Remove this title and section

File: software-containers/arcade-casa/README.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+If you have previously created a desktop session, you can use the UI of that session to access your container. There is an 'Applications' tab on the top left hand corner. Select 'Applications->CANFAR->CASA' and then the version under development. An xterm for the selected CASA version will pop up after some time.
+#### Performing Tests
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]: Remove this title

File: software-containers/arcade-casa/README.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+The xterm provides access to the CASA container under development. Perform the necessary tests and repeat the build, deploy and test process if necessary. The following is an example test session on the xterm.
```
Comment (brianmajor) [inline software-containers/arcade-casa/README.md]:
```suggestion
The following steps can be performed in a graphical container environment (such as arcade) to test that the container was built correctly.
```

**Review by brianmajor — CHANGES_REQUESTED** [review]
Alinga, along with the comments, could you please change the test steps to be just a list of steps rather than terminal output?  For example:

1.  Type the `id` command and ensure that user name is displayed
2.  Run the `casa` command and ensure the logging window opens.

etc...

File: software-containers/arcade-casa/README.md (hunk 11 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Under some circumstances, when creating a new version of CASA, a new folder would need to be created. In that case, we would need to create a Makefile and a Dockerfile. For some updates, we may need to update the Dockerfile and/or the download.sh file. For a new issue fix, we may need to create a file to build and install the fix onto the new container.

Comment (brianmajor) [reply software-containers/arcade-casa/README.md]: The documentation should only be about building what's currently in git.

File: software-containers/arcade-casa/README.md (hunk 10 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Done

File: software-containers/arcade-casa/README.md (hunk 11 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md:4]: For some updates, we may need to update the download.sh file.

Comment (yeunga) [reply software-containers/arcade-casa/README.md:4]: For a new issue fix, we may need to create a file and place it in git.

Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Deleted the step.

File: software-containers/arcade-casa/README.md (hunk 15 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Thanks. Done.

File: software-containers/arcade-casa/README.md (hunk 14 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Done.

File: software-containers/arcade-casa/README.md (hunk 13 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Done

File: software-containers/arcade-casa/README.md (hunk 12 above)
Comment (yeunga) [reply software-containers/arcade-casa/README.md]: Done.

**Review by brianmajor — CHANGES_REQUESTED** [review]
Looks like there are some markdown formatting errors.  You can use the 'preview option' to iterate on it from your fork.

**Review by brianmajor — APPROVED** [review]

**Pull request merged**